Objects/typeobject.c

```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
             name, type->tp_name);
         return -1;
     }
-    if (PyUnicode_Check(name)) {
-        if (PyUnicode_CheckExact(name)) {
-            Py_INCREF(name);
+    if (!PyUnicode_Check(name)) {
+        PyErr_Format(PyExc_TypeError,
+                     "attribute name must be string, not '%.200s'",
+                     Py_TYPE(name)->tp_name);
+        return -1;
+    }
+
+    if (PyUnicode_CheckExact(name)) {
+        Py_INCREF(name);
+    }
+    else {
+        name = _PyUnicode_Copy(name);
+        if (name == NULL)
+            return -1;
+    }
+    /* bpo-40521: Interned strings are shared by all subinterpreters */
+    if (!PyUnicode_CHECK_INTERNED(name)) {
+        PyUnicode_InternInPlace(&name);
+        if (!PyUnicode_CHECK_INTERNED(name)) {
+            PyErr_SetString(PyExc_MemoryError,
+                            "Out of memory interning an attribute name");
+            Py_DECREF(name);
+            return -1;
         }
-        else {
-            name = _PyUnicode_Copy(name);
-            if (name == NULL)
-                return -1;
+    }
+
+    PyTypeObject *metatype = Py_TYPE(type);
+    assert(!_PyType_HasFeature(metatype, Py_TPFLAGS_INLINE_VALUES));
+    assert(!_PyType_HasFeature(metatype, Py_TPFLAGS_MANAGED_DICT));
+
+    PyObject *old_value;
+    PyObject *descr = _PyType_Fetch(metatype, name);
+    if (descr != NULL) {
+        descrsetfunc f = Py_TYPE(descr)->tp_descr_set;
+        if (f != NULL) {
+            old_value = NULL;
+            res = f(descr, (PyObject *)type, value);
+            goto done;
         }
-        /* bpo-40521: Interned strings are shared by all subinterpreters */
-        if (!PyUnicode_CHECK_INTERNED(name)) {
-            PyUnicode_InternInPlace(&name);
-            if (!PyUnicode_CHECK_INTERNED(name)) {
-                PyErr_SetString(PyExc_MemoryError,
-                                "Out of memory interning an attribute name");
-                Py_DECREF(name);
-                return -1;
-            }
+    }
+
+    PyObject *dict = type->tp_dict;
+    if (dict == NULL) {
+        // We don't just do PyType_Ready because we could already be readying
+        BEGIN_TYPE_LOCK();
+        dict = type->tp_dict;
+        if (dict == NULL) {
+            dict = type->tp_dict = PyDict_New();
         }
+        END_TYPE_LOCK();
     }
-    else {
-        /* Will fail in _PyObject_GenericSetAttrWithDict. */
-        Py_INCREF(name);
+
+    // We don't want any re-entrancy between when we update the dict
+    // and call type_modified_unlocked, including running the destructor
+    // of the current value as it can observe the cache in an inconsistent
+    // state.  Because we have an exact unicode and our dict has exact
+    // unicodes we know that this will all complete without releasing
+    // the locks.
+    BEGIN_TYPE_DICT_LOCK(dict);
+
+    if (_PyDict_GetItemRef_LockHeld(dict, name, &old_value) < 0) {
```

Comment:
We're doing both "preserve the old value" and "invalidate the cache early" strategies. Is preserving the old value necessary for safety?

(To be clear: I'm not opposed to this; I'm just trying to understand the implementation and constraints).

Reply:
This is necessary to prevent any re-entrancy during the update (when the existing value is replaced and its finalizer runs). The problem with clearing the version tag but not clearing `Py_TPFLAGS_VALID_VERSION_TAG` is that we're in a pretty funky state.  If we start running Python code again and hit `assign_version_tag` anywhere we'll happily think we have a version tag assigned even though our version is invalid.  That means we can do a lookup against that and get a cache hit.  Ultimately that breaks in the new test in `test_class`.

By guaranteeing that we have no re-entrancy we're not exposing this funky intermediate state.  And we need to have this funky intermediate state so that we can both not have any cache hits as well as being able to know what types we need to clear later when we do the full `type_modified_unlocked`.

The alternative here I think would be to clear both the type version tag and the has version flag.  That would also require your originally proposed flag to disable the type cache.  But the challenging thing there is then finishing the 2nd half of the `type_modified_unlocked` - we need to remember all of the types that we've cleared, but we lost the state that's tracking that (`Py_TPFLAGS_VALID_VERSION_TAG`).  So that would require either another flag to say we're in the process of clearing the tags or we'd need to store all of the types that we're clearing in a list.

But if we avoid the re-entrancy then everything's fine, and assigning a new version number to the types is protected by the type lock.

Reply:
Oh, also I wanted to assert that we're never caching w/ a 0 version tag.  And that works fine everywhere except for in some tests for specialization which are using `_testcapi` to assign a 0 version tag and say it's valid :(